### PR TITLE
435 data functions should be made optional

### LINKED ISSUE #435: Data functions should be made optional
Instead of returning a bad null value (like -1 degrees or -1 as an apriltag id), make the return type an optional of the current type.
This will have a large impact as it affects DragonLimelight, DragonPhotonCam, and DragonVision.
The following methods should be made optional if we have no target, cannot access the network table, or can't get data for any other reason:
GetTargetYaw
GetTargetYawRobotFrame
GetTargetPitchRobotFrame
GetTargetPitch
GetPipelineLatency
GetTargetSkew
GetTargetArea
GetAprilTagID
EstimateTargetXDistance
EstimateTargetYDistance
EstimateTargetZDistance
EstimateTargetXDistance_RelToRobotCoords
EstimateTargetYDistance_RelToRobotCoords
EstimateTargetZDistance_RelToRobotCoords

### PR DESCRIPTION
**Pull request by ImAPenguin56**
Nick better fix this
Closes #435

### CODE REVIEW

**Review by nickfrisch — CHANGES_REQUESTED** [review]
If we don't have good data, need to return std::nullopt, also need to wrap all of our valid retursn in std::make_optional()

File: src/main/cpp/DragonVision/DragonLimelight.cpp [hunk 1]
```diff
@@ -437,54 +449,53 @@ units::length::inch_t DragonLimelight::EstimateTargetZDistance()
...
+std::optional<units::length::inch_t> DragonLimelight::EstimateTargetXDistance_RelToRobotCoords()
 {
-    if (EstimateTargetXDistance().to<double>() != -1.0)
+    if (EstimateTargetXDistance())
     {
-        units::length::inch_t targetXoffset_RF_inch = EstimateTargetXDistance() + GetMountingXOffset(); ///< the offset is negative if the limelight is behind the center of the robot
+        units::length::inch_t targetXoffset_RF_inch = EstimateTargetXDistance().value() + GetMountingXOffset(); ///< the offset is negative if the limelight is behind the center of the robot
 
         return targetXoffset_RF_inch;
     }
```
Comment (nickfrisch) [inline src/main/cpp/DragonVision/DragonLimelight.cpp:459]: Need to return std::nullopt

File: src/main/cpp/DragonVision/DragonLimelight.h [hunk 2]
```diff
@@ -137,6 +138,4 @@ class DragonLimelight : public DragonCamera
     units::angle::degree_t GetTy() const;
```
Comment (nickfrisch) [inline src/main/cpp/DragonVision/DragonLimelight.h:138]: GetTx and GetTy should be optional

File: src/main/cpp/DragonVision/DragonPhotonCam.cpp [hunk 3]
```diff
@@ -280,7 +278,7 @@ units::time::millisecond_t DragonPhotonCam::GetPipelineLatency()
     return latency;
```
Comment (nickfrisch) [inline src/main/cpp/DragonVision/DragonPhotonCam.cpp:278]: need to return nullopt if we don't have a latest result

File: src/main/cpp/DragonVision/DragonVision.cpp [hunk 4]
```diff
@@ -164,33 +164,32 @@ std::optional<VisionData> DragonVision::GetVisionDataToNearestStageTag(VISION_EL
...
+		if (launcherTagId)
 			selectedCam = m_dragonCameraMap[RobotElementNames::CAMERA_USAGE::LAUNCHER];
 		else
 			selectedCam = m_dragonCameraMap[RobotElementNames::CAMERA_USAGE::PLACER];
 	}
 
 	if (selectedCam != nullptr)
 	{
-		return selectedCam->GetDataToNearestAprilTag();
+		return selectedCam->GetDataToNearestAprilTag().value();
```
Comment (nickfrisch) [inline src/main/cpp/DragonVision/DragonVision.cpp:192]: Need to check if we have valid data

File: src/main/cpp/DragonVision/DragonVision.cpp [hunk 5]
```diff
@@ -200,7 +199,8 @@ std::optional<VisionData> DragonVision::GetDataToNearestAprilTag(RobotElementNam
 {
 	if (m_dragonCameraMap[position] != nullptr)
 	{
-		return m_dragonCameraMap[position]->GetDataToNearestAprilTag();
+
+		return m_dragonCameraMap[position]->GetDataToNearestAprilTag().value();
```
Comment (nickfrisch) [inline src/main/cpp/DragonVision/DragonVision.cpp:203]: Need to check if data is valid

File: src/main/cpp/DragonVision/DragonVision.cpp [hunk 6]
```diff
@@ -229,8 +229,8 @@ std::optional<VisionData> DragonVision::GetVisionDataFromNote(VISION_ELEMENT ele
 		else if (frontHasDetection && backHasDetection)
 		{
 			// check which note is closest to robot
-			frc::Translation2d translationLauncher = frc::Translation2d(m_dragonCameraMap[RobotElementNames::CAMERA_USAGE::LAUNCHER_INTAKE]->EstimateTargetXDistance_RelToRobotCoords(), m_dragonCameraMap[RobotElementNames::CAMERA_USAGE::LAUNCHER_INTAKE]->EstimateTargetYDistance_RelToRobotCoords());
-			frc::Translation2d translationPlacer = frc::Translation2d(m_dragonCameraMap[RobotElementNames::CAMERA_USAGE::PLACER_INTAKE]->EstimateTargetXDistance_RelToRobotCoords(), m_dragonCameraMap[RobotElementNames::CAMERA_USAGE::PLACER_INTAKE]->EstimateTargetYDistance_RelToRobotCoords());
+			frc::Translation2d translationLauncher = frc::Translation2d(m_dragonCameraMap[RobotElementNames::CAMERA_USAGE::LAUNCHER_INTAKE]->EstimateTargetXDistance_RelToRobotCoords().value(), m_dragonCameraMap[RobotElementNames::CAMERA_USAGE::LAUNCHER_INTAKE]->EstimateTargetYDistance_RelToRobotCoords().value());
```
Comment (nickfrisch) [inline src/main/cpp/DragonVision/DragonVision.cpp:232]: Check if EstimateDistances have valid values

File: src/main/cpp/DragonVision/DragonVision.cpp [hunk 7]
```diff
@@ -252,10 +252,10 @@ std::optional<VisionData> DragonVision::GetVisionDataFromNote(VISION_ELEMENT ele
 	if (selectedCam != nullptr)
 	{
 		// create translation using 3 estimated distances
-		frc::Translation3d translationToNote = frc::Translation3d(selectedCam->EstimateTargetXDistance_RelToRobotCoords(), selectedCam->EstimateTargetYDistance_RelToRobotCoords(), selectedCam->EstimateTargetZDistance_RelToRobotCoords());
+		frc::Translation3d translationToNote = frc::Translation3d(selectedCam->EstimateTargetXDistance_RelToRobotCoords().value(), selectedCam->EstimateTargetYDistance_RelToRobotCoords().value(), selectedCam->EstimateTargetZDistance_RelToRobotCoords().value());
```
Comment (nickfrisch) [inline src/main/cpp/DragonVision/DragonVision.cpp:255]: Check if EstimateDistances have valid data

File: src/main/cpp/DragonVision/DragonVision.cpp [hunk 8]
```diff
@@ -269,14 +269,14 @@ std::optional<VisionData> DragonVision::GetVisionDataFromElement(VISION_ELEMENT
 {
 	DragonCamera *selectedCam = nullptr;
 
-	int launcherTagId = m_dragonCameraMap[RobotElementNames::CAMERA_USAGE::LAUNCHER]->GetAprilTagID();
-	int placerTagId = m_dragonCameraMap[RobotElementNames::CAMERA_USAGE::PLACER]->GetAprilTagID();
+	int launcherTagId = m_dragonCameraMap[RobotElementNames::CAMERA_USAGE::LAUNCHER]->GetAprilTagID().value();
```
Comment (nickfrisch) [inline src/main/cpp/DragonVision/DragonVision.cpp:272]: Make sure GetAprilTagId has valid data

File: src/main/cpp/DragonVision/DragonVision.cpp [hunk 9]
```diff
@@ -314,7 +314,7 @@ std::optional<VisionData> DragonVision::GetVisionDataFromElement(VISION_ELEMENT
 	}
 
 	// optional of the April Tag's 3D pose
-	std::optional<frc::Pose3d> optionalAprilTagPose = GetAprilTagLayout().GetTagPose(selectedCam->GetAprilTagID());
+	std::optional<frc::Pose3d> optionalAprilTagPose = GetAprilTagLayout().GetTagPose(selectedCam->GetAprilTagID().value());
```
Comment (nickfrisch) [inline src/main/cpp/DragonVision/DragonVision.cpp:317]: Make sure GetAprilTagId has valid data

**Review by nickfrisch — CHANGES_REQUESTED** [review]

File: src/main/cpp/DragonVision/DragonLimelight.cpp [hunk 10]
```diff
@@ -195,7 +196,7 @@ units::angle::degree_t DragonLimelight::GetTy() const
     return units::angle::degree_t(0.0);
```
[the remark this answers is not in the document]
Comment (ImAPenguin56) [reply src/main/cpp/DragonVision/DragonLimelight.cpp:196]: not made optional

File: src/main/cpp/DragonVision/DragonLimelight.cpp [hunk 11]
```diff
@@ -195,7 +196,7 @@ units::angle::degree_t DragonLimelight::GetTy() const
...
 }
 
-units::angle::degree_t DragonLimelight::GetTargetYaw()
+std::optional<units::angle::degree_t> DragonLimelight::GetTargetYaw()
```
[the remark this answers is not in the document]
Comment (ImAPenguin56) [reply src/main/cpp/DragonVision/DragonLimelight.cpp:199]: GetTX cannot be nullopt

File: src/main/cpp/DragonVision/DragonLimelight.cpp [hunk 12]
```diff
@@ -217,23 +218,26 @@ units::angle::degree_t DragonLimelight::GetTargetYaw()
...
-    return angleOffset;
+        units::angle::radian_t angleOffset = units::angle::radian_t(atan((targetHorizOffsetRobotFrame / targetDistanceRobotFrame).to<double>()));
+        return angleOffset;
+    }
+    else
+        return std::nullopt;
 }
 
-units::angle::degree_t DragonLimelight::GetTargetPitch()
+std::optional<units::angle::degree_t> DragonLimelight::GetTargetPitch()
```
[the remark this answers is not in the document]
Comment (ImAPenguin56) [reply src/main/cpp/DragonVision/DragonLimelight.cpp:240]: both functions cannot be opt

File: src/main/cpp/DragonVision/DragonLimelight.cpp (hunk 1 above)
Comment (ImAPenguin56) [reply src/main/cpp/DragonVision/DragonLimelight.cpp:459]: done

File: src/main/cpp/DragonVision/DragonVision.cpp (hunk 5 above)
Comment (ImAPenguin56) [reply src/main/cpp/DragonVision/DragonVision.cpp:203]: done

File: src/main/cpp/DragonVision/DragonVision.cpp (hunk 7 above)
Comment (ImAPenguin56) [reply src/main/cpp/DragonVision/DragonVision.cpp:255]: unnecessary

File: src/main/cpp/DragonVision/DragonVision.cpp (hunk 8 above)
Comment (ImAPenguin56) [reply src/main/cpp/DragonVision/DragonVision.cpp:272]: done

File: src/main/cpp/DragonVision/DragonVision.cpp (hunk 9 above)
Comment (ImAPenguin56) [reply src/main/cpp/DragonVision/DragonVision.cpp:317]: already there

File: src/main/cpp/DragonVision/DragonLimelight.cpp [hunk 13]
```diff
@@ -217,23 +218,26 @@ units::angle::degree_t DragonLimelight::GetTargetYaw()
...
-units::angle::degree_t DragonLimelight::GetTargetYawRobotFrame()
+std::optional<units::angle::degree_t> DragonLimelight::GetTargetYawRobotFrame()
 {
     // Get the horizontal angle to the target and convert to radians
-    units::angle::radian_t limelightFrameHorizAngleRad = GetTargetYaw();
-
-    units::length::inch_t targetXdistance = EstimateTargetXDistance();
-
-    units::length::inch_t targetHorizOffset = targetXdistance * tan(limelightFrameHorizAngleRad.to<double>());
+    std::optional<units::angle::radian_t> limelightFrameHorizAngleRad = GetTargetYaw();
```
[the remark this answers is not in the document]
Comment (ImAPenguin56) [reply src/main/cpp/DragonVision/DragonLimelight.cpp:224]: already did that

**Review by nickfrisch — CHANGES_REQUESTED** [review]
Fix some std::optional stuff

File: src/main/cpp/DragonVision/DragonLimelight.cpp [hunk 14]
```diff
@@ -70,15 +70,16 @@ DragonLimelight::DragonLimelight(
 
 /// @brief Assume that the current pipeline is AprilTag and that a target is detected
 /// @return -1 if the network table cannot be found
-int DragonLimelight::GetAprilTagID()
+std::optional<int> DragonLimelight::GetAprilTagID()
```
Comment (reslip1) [inline src/main/cpp/DragonVision/DragonLimelight.cpp:73]: I didn't think we were using limelights for april tag this year.

Comment (nickfrisch) [reply src/main/cpp/DragonVision/DragonLimelight.cpp:73]: We're not, but we had the functionality and this needed to be consistent with PhotonCamera

File: src/main/cpp/DragonVision/DragonLimelight.cpp [hunk 15]
```diff
@@ -70,15 +70,16 @@ DragonLimelight::DragonLimelight(
...
 {
     auto nt = m_networktable.get();
     if (nt != nullptr)
     {
-        return round(nt->GetNumber("tid", -1));
+        int aprilTagInt = int(round(nt->GetNumber("tid", -1)));
```
Comment (muscatd75) [inline src/main/cpp/DragonVision/DragonLimelight.cpp:78]: please change to something like:
double value = nt->GetNumber("tid", -1);
int aprilTagInt = static_cast<int>(value + (value > 0 ? 0.5 : -0.5 ))

File: src/main/cpp/DragonVision/DragonLimelight.cpp [hunk 16]
```diff
@@ -255,49 +259,52 @@ units::angle::degree_t DragonLimelight::GetTargetPitch()
...
-double DragonLimelight::GetTargetArea()
+std::optional<double> DragonLimelight::GetTargetArea()
 {
     auto nt = m_networktable.get();
     if (nt != nullptr)
     {
         return nt->GetNumber("ta", 0.0);
     }
-    return 0.0;
+    else
```
Comment (muscatd75) [inline src/main/cpp/DragonVision/DragonLimelight.cpp:284]: you do not need "else"

File: src/main/cpp/DragonVision/DragonLimelight.cpp [hunk 17]
```diff
@@ -255,49 +259,52 @@ units::angle::degree_t DragonLimelight::GetTargetPitch()
...
-units::time::millisecond_t DragonLimelight::GetPipelineLatency()
+std::optional<units::time::millisecond_t> DragonLimelight::GetPipelineLatency()
 {
     auto nt = m_networktable.get();
     if (nt != nullptr)
     {
         return units::time::second_t(nt->GetNumber("tl", 0.0));
     }
-    return units::time::second_t(0.0);
+    else
```
Comment (muscatd75) [inline src/main/cpp/DragonVision/DragonLimelight.cpp:306]: no need for else

File: src/main/cpp/DragonVision/DragonLimelight.cpp [hunk 18]
```diff
@@ -380,16 +387,18 @@ void DragonLimelight::PrintValues()
...
-
-    if (GetAprilTagID() == -1)
+    std::optional<units::angle::degree_t> targetPitch = GetTargetPitch();
+    std::optional<int>
+        aprilTagID = GetAprilTagID();
+    if (aprilTagID && targetPitch)
     {
         // d=(h2-h1)/tan(a1+a2)
-        units::length::inch_t estimatedTargetDistance = (m_noteVerticalOffset - mountingHeight) / units::math::tan(mountingAngle + GetTargetPitch());
+        units::length::inch_t estimatedTargetDistance = (m_noteVerticalOffset - mountingHeight) / units::math::tan(mountingAngle + targetPitch.value());
```
Comment (muscatd75) [inline src/main/cpp/DragonVision/DragonLimelight.cpp:401]: what if tan returns 0?

File: src/main/cpp/DragonVision/DragonVision.cpp [hunk 19]
```diff
@@ -198,9 +197,10 @@ std::optional<VisionData> DragonVision::GetVisionDataToNearestTag()
 
 std::optional<VisionData> DragonVision::GetDataToNearestAprilTag(RobotElementNames::CAMERA_USAGE position)
 {
-	if (m_dragonCameraMap[position] != nullptr)
+	std::optional<VisionData> dataToAprilTag = m_dragonCameraMap[position]->GetDataToNearestAprilTag();
+	if (m_dragonCameraMap[position] != nullptr && dataToAprilTag.has_value())
```
Comment (muscatd75) [inline src/main/cpp/DragonVision/DragonVision.cpp:201]: add parenthesis

File: src/main/cpp/DragonVision/DragonVision.cpp [hunk 20]
```diff
@@ -220,23 +220,23 @@ std::optional<VisionData> DragonVision::GetVisionDataFromNote(VISION_ELEMENT ele
...
-		bool backHasDetection = m_dragonCameraMap[RobotElementNames::CAMERA_USAGE::P_INTAKE]->HasTarget();
-		if (!frontHasDetection && !backHasDetection)
+		bool launcherHasDetection = m_dragonCameraMap[RobotElementNames::CAMERA_USAGE::L_INTAKE]->HasTarget();
+		bool placerHasDetection = m_dragonCameraMap[RobotElementNames::CAMERA_USAGE::P_INTAKE]->HasTarget();
+		if (!launcherHasDetection && !placerHasDetection)
 		{
 			return std::nullopt;
 		}
-		else if (frontHasDetection && backHasDetection)
+		else if (launcherHasDetection && placerHasDetection) // we see targets in no cameras
```
Comment (muscatd75) [inline src/main/cpp/DragonVision/DragonVision.cpp:229]: the comment seems wrong. The if is asking if both launcher and placer see a target

**Review by muscatd75 — CHANGES_REQUESTED** [review]
See comments

File: src/main/cpp/DragonVision/DragonLimelight.cpp (hunk 15 above)
Comment (nickfrisch) [reply src/main/cpp/DragonVision/DragonLimelight.cpp:78]: Done

File: src/main/cpp/DragonVision/DragonLimelight.cpp [hunk 21]
```diff
@@ -217,23 +218,26 @@ units::angle::degree_t DragonLimelight::GetTargetYaw()
...
+        units::length::inch_t targetHorizOffset = targetXdistance.value() * tan(limelightFrameHorizAngleRad.value().to<double>());
 
-    units::length::inch_t targetHorizOffsetRobotFrame = targetHorizOffset + GetMountingYOffset(); // the offset is positive if the limelight is to the left of the center of the robot
-    units::length::inch_t targetDistanceRobotFrame = targetXdistance + GetMountingXOffset();      // the offset is negative if the limelight is behind the center of the robot
+        units::length::inch_t targetHorizOffsetRobotFrame = targetHorizOffset + GetMountingYOffset();    // the offset is positive if the limelight is to the left of the center of the robot
+        units::length::inch_t targetDistanceRobotFrame = targetXdistance.value() + GetMountingXOffset(); // the offset is negative if the limelight is behind the center of the robot
 
-    units::angle::radian_t angleOffset = units::angle::radian_t(atan((targetHorizOffsetRobotFrame / targetDistanceRobotFrame).to<double>()));
-    return angleOffset;
+        units::angle::radian_t angleOffset = units::angle::radian_t(atan((targetHorizOffsetRobotFrame / targetDistanceRobotFrame).to<double>()));
```
[the remark this answers is not in the document]
Comment (nickfrisch) [reply src/main/cpp/DragonVision/DragonLimelight.cpp]: Switched to atan2

File: src/main/cpp/DragonVision/DragonLimelight.cpp (hunk 17 above)
Comment (nickfrisch) [reply src/main/cpp/DragonVision/DragonLimelight.cpp:306]: Done

File: src/main/cpp/DragonVision/DragonLimelight.cpp (hunk 18 above)
Comment (nickfrisch) [reply src/main/cpp/DragonVision/DragonLimelight.cpp:401]: Fixed

File: src/main/cpp/DragonVision/DragonVision.cpp (hunk 19 above)
Comment (nickfrisch) [reply src/main/cpp/DragonVision/DragonVision.cpp:201]: Done

File: src/main/cpp/DragonVision/DragonVision.cpp (hunk 20 above)
Comment (nickfrisch) [reply src/main/cpp/DragonVision/DragonVision.cpp:229]: Done

**Review by nickfrisch — APPROVED** [review]

**Comment by witcpalek** [comment]
Please update branch / merge with main

**Comment by nickfrisch** [comment]
> Please update branch / merge with main

@witcpalek Updated

**Review by muscatd75 — APPROVED** [review]

**Pull request merged**